### PR TITLE
Create tool to document the database

### PR DESCRIPTION
**Pull request by mlazaric**
## Introduction

I think one of the harder things to understand about the code base is the database. There are some columns which are not intuitive or are ambiguous, some tables are hard to find etc. Currently the easiest solution is to search the code base for its usage and extrapolate the logic from that or login to the test server/VM and explore the existing data there. So to combat that, I think we should document the database structure and its semantics and to simplify that process I made a simple PHP script to generate (**and maintain**) the database documentation files.

## Documentation file structure

Here is an example documentation file generated from the `wordcheck_events` table: https://gist.github.com/mlazaric/c696738b74817c8aa9d0d7a3880d8b9e

The file has two parts: the table of columns and the column definitions. The table of columns contains some basic information about each column in the table, such as its name, type, default value etc. The table is autogenerated and should not be manually edited since any changes will be overwritten by the script. The column definitions (`## COLUMN_NAME`) are used to explain a column and its usage if necessary, for example it could explain the format of the suggestions and corrections columns. Anything else (lines that are not part of the table nor are they prefixed with `## `) are ignored by the script and can be manually edited.

Here is the same file filled in with some column descriptions: https://gist.github.com/mlazaric/c462f32b3c1b09a62d7ef12b4f776f64

We do not necessarily have to describe each column since some of them, like the `timestamp` or `check_id`, are self explanatory. But we do have to have a column definition for each column. 

## Supported operations

The script will support creating the documentation files, updating them with new information and checking whether they are up to date with the database. The checking operation would be used in CI to ensure the documentation does not go stale and the other two would simplify maintenance. 

### Generate

If the second argument is all, it generates a documentation file for each table in the database. Otherwise, it uses the second argument as the table name and generates the documentation file for that table.

#### Syntax

```
php document_database.php generate <directory path> <table name or all>
```

### Update 

If the second argument is all, it updates the documentation file for each table in the database. Otherwise, it uses the second argument as the table name and updates the documentation file for that table.

Updating consists of regenerating the table of columns, adding new column definitions if any columns were added to the table and deleting column definitions of columns that no longer exist in the table.

#### Syntax

```
php document_database.php update <directory path> <table name or all>
```

### Verify 

If the second argument is all, it checks the documentation file for each table in the database. Otherwise, it uses the second argument as the table name and checks the documentation file for that table.

Checking consists of the following conditions:
1. The table of columns in the documentation file is equal to the table generated with the current information from the database. This ensures any changes such as changing the default value of a column are updated in the documentation.
2. The documentation file contains a column definition for every column in the table. This ensures any new columns are added to the documentation.
3. The documentation file does not contain a column definition for a column that does not exist in the table. This ensures any column deletion is documented properly. 

If any of those three conditions are not satisfied, the script will exit with an exit code of 1 and print out the failed conditions. This operation can be used in CI to ensure we document any database changes.  

#### Syntax

```
php document_database.php verify <directory path> <table name or all>
```

## Miscellaneous

 - Project tables are special cased, since it does not make sense to document each instance of a project table separately, they are documented as the `$projectid` table.

### CODE REVIEW

**Comment by cpeel** [comment]
I love the proposal and like where you are going with it.

I'm not a fan of the hoops you are jumping through to allow unit testing of the file. Consider instead creating a class in `pinc/` that encapsulates the doc creation logic and write your tests against that. It's frustrating that php doesn't offer proper mocking support like python does, but you might also find phpunit's [test doubles](https://phpunit.readthedocs.io/en/7.0/test-doubles.html) useful.

File: SETUP/document_database.php [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        echo "File path '$directory_path' does not exist or is not a directory.\n";
+
+        exit(1);
+    }
+
+    if ($table_name === 'all') {
+        generate_files_for_all_tables($directory_path);
+    }
+    else {
+        generate_file_for_table($table_name, $directory_path . '/' . $table_name . '.md', $table_name);
```
Comment (cpeel) [inline SETUP/document_database.php]: When creating filenames and paths, the rest of the code (or at least the bulk of it) does it with string interpolation, eg:
```
        generate_file_for_table($table_name, "$directory_path/$table_name.md", $table_name);
```
Lets continue to do that in this file too please.

File: SETUP/document_database.php [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    foreach ($table_names as $row) {
+        $table_name = $row[0];
+
+        if (startswith($table_name, 'projectID')) {
+            if ($was_projectid_table_generated) continue;
+
+            $was_projectid_table_generated = true;
+
+            $operation($table_name, $directory_path . '/$projectid.md', '$projectid');
```
Comment (cpeel) [inline SETUP/document_database.php]:
```
$operation($table_name, "$directory_path/\$projectid.md", '$projectid');
```

File: SETUP/document_database.php [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        }
+        else {
+            $operation($table_name, $directory_path . '/' . $table_name . '.md', $table_name);
```
Comment (cpeel) [inline SETUP/document_database.php]:
```
$operation($table_name, "$directory_path/$table_name.md", $table_name);
```

**Comment by mlazaric** [comment]
I agree on both the design being better and easier to test and using `projectIDxxxxxxxxxxxxx`. :)

**Comment by srjfoo** [comment]
>> Awesome, I think this design makes more sense.
>> 
>> I do wonder if we should rename the generic "project" table from `$projectid` to something else. The `$` is a PITA to escape and probably also to deal with in the filesystem. Perhaps `projectIDxxxxxxxxxxxxx`?
>
> I agree on both the design being better and easier to test and using `projectIDxxxxxxxxxxxxx`. :)

Assumption: We only care about the most recent version of the projectid tables? I have recently unarchived projects from the original 2-round system. There are still tables with 21 fields instead of 23. And there are multiple versions of the metadata set.

File: SETUP/document_database.php [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+// generate <directory_path> <table_name or all>
+if ($argv[1] === 'generate') {
+    if ($argc !== 4) {
+        echo "Operation generate requires 2 arguments, $number_of_arguments were given.\n";
+        echo "Supported syntax for generate command is 'generate <directory path> <table name or all>'.\n";
+
+        exit(1);
+    }
+
+    $directory_path = $argv[2];
```
Comment (cpeel) [inline SETUP/document_database.php]: 99% of the time we will want to run this against a directory that exists in the code checkout. It would be useful to have this default to that directory. Perhaps swap this and the `table` arguments so that `directory_path` is last.

Consider a directory name of `SETUP/dbdocs/` (so `$relPath/../SETUP/dbdocs/`).

File: SETUP/document_database.php [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    $rows = mysqli_fetch_all($result, MYSQLI_ASSOC);
+    mysqli_free_result($result);
+
+    return $rows;
+}
+
+/**
+ * Runs a "SHOW TABLES" query and returns the result.
+ */
+function query_table_names_in_current_database(): array {
```
[the remark this answers is not in the document]
Comment (mlazaric) [reply SETUP/document_database.php]: One problem with this approach is that not all tables are created in `db_schema.sql`, the best (and currently only?) example are the project id tables. Although I guess we will have to special case those in any case, so I agree with this solution. 

On a related note, are any project id tables created during CI? That would be useful for when we add the verification part to the CI.

**Comment by mlazaric** [comment]
Those two steps sound good to me. :+1:

File: SETUP/document_database.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (cpeel) [reply SETUP/document_database.php]: `projectID` tables should be the only things created in the code. The CI doesn't (yet?) create any of them but it should be easy to call `project_allow_pages()` somewhere so that one is created.

File: SETUP/document_database.php (hunk 4 above)
Comment (mlazaric) [reply SETUP/document_database.php]: Done in c07c32a.

File: SETUP/document_database.php [hunk 6]
```diff
@@ -0,0 +1,273 @@
+<?php
```
[the remark this answers is not in the document]
Comment (mlazaric) [reply SETUP/document_database.php]: Done in c07c32a.

File: SETUP/document_database.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (mlazaric) [reply SETUP/document_database.php]: To get the table name of a "live" project id table, we could: 

 * Create a table `projectIDxxxxxxxxxxxxx` in both CI and on the test server (with `project_allow_pages`) and then hardcode that table name in the function. 
 * Query the database for the most recently updated project id table name and use in the function.

[the remark this answers is not in the document]
Comment (srjfoo) [reply SETUP/document_database.php]: I think we're going to have to special case the `projectID` table description, regardless. Do we absolutely have to create a table to query, or use the most recently created `projectID` table? What I'm asking is, can we not just work with the information in `project_allow_pages`? If that can't be used, then the possible issue I see with querying the database for the most recently created `projectID` table is that on systems like the test server, there's the possibility that `project_allow_pages` might, in edge cases, have been updated since the most recent project creation.

[the remark this answers is not in the document]
Comment (mlazaric) [reply SETUP/document_database.php]: > What I'm asking is, can we not just work with the information in `project_allow_pages`?

Hmm, that is a good question, we _can_ parse the information from the `CREATE TABLE` query, however it is simpler just to query the database for that information. I think parsing the query would be more complex to implement and maintain. 

> If that can't be used, then the possible issue I see with querying the database for the most recently created `projectID` table is that on systems like the test server, there's the possibility that `project_allow_pages` might, in edge cases, have been updated since the most recent project creation.

Ah, then I think the best solution is to manually create and maintain (whenever necessary) a `projectIDxxxxxxxxxxxxx` table. It is more explicit and will be easier to debug.

[the remark this answers is not in the document]
Comment (srjfoo) [reply SETUP/document_database.php]: > Ah, then I think the best solution is to manually create and maintain (whenever necessary) a `projectIDxxxxxxxxxxxxx` table. It is more explicit and will be easier to debug.

"Create and maintain"? Or "Create and delete when the update is done"? I.e. do you really want a `projectID` table hanging around that really doesn't have anything to do with anything? It shouldn't have an entry in the `projects` table, and I would think that things would be cleaner if it were simply deleted. That way you don't have to worry about it creating problems anywhere else in the code.

[the remark this answers is not in the document]
Comment (mlazaric) [reply SETUP/document_database.php]: We can create it before running the command and delete it afterwards. By maintain I just meant recreating it if we ever change the create table query in code. :)

[the remark this answers is not in the document]
Comment (cpeel) [reply SETUP/document_database.php]: I agree with this approach. We modify projectID table structures very very rarely.

[the remark this answers is not in the document]
Comment (mlazaric) [reply SETUP/document_database.php]: Resolved in c3b940f9214b4d9e0ce7909544b1700859bbfd4d.

File: SETUP/tests/phpunit_bootstrap.php [hunk 7]
```diff
@@ -4,3 +4,4 @@
 include_once($relPath.'base.inc');
 include_once($relPath.'misc.inc');
 include_once($relPath.'wordcheck_engine.inc');
+include_once($relPath.'TableDocumentation.inc');
```
Comment (cpeel) [inline SETUP/tests/phpunit_bootstrap.php:7]: Technically you don't need this, because the `dp_class_autoloader` will automagically load files for classes where the class name matches the filename. This is fine, but just FYI.

**Comment by cpeel** [comment]
I have a few small updates that I'd like to get into this MR -- they're the last three commits in https://github.com/cpeel/dproofreaders/commits/database-documentation. If they look good to you can cherry-pick them into your branch (or squash them into others as you see fit).

**Comment by mlazaric** [comment]
Thanks for the fixes! Looks great, fast forwarded them onto this branch.

File: SETUP/tests/phpunit_bootstrap.php (hunk 7 above)
Comment (mlazaric) [reply SETUP/tests/phpunit_bootstrap.php:7]: Ah, that's useful, thanks for the info!
